### ISSUE TITLE
API to access desired periods of timesets?

### ISSUE THREAD

**Issue by coreyeng** [Origen-SDK/origen#307]
Hello,

I've been continuing to work on running direct text source though OrigenSim (from either <code>.atp</code> or <code>.avc</code> source) and have run into something that I think could use an 'official solution', if it doesn't already have one that I'm just missing.

I can set up all the <code>timesets</code> and query them from Origen fine, which allows the simulator to drive, compare, etc. as expected. When these are set, this is usually one of the first thing that happens in <code>dut.startup</code>, but, when running a direct source through the simulator,  I'm bypassing <code>startup</code> (and any other non-setup code), leaving out any invocations to <code>set_timeset</code> and leaving it up to me to set the timeset based on what the pattern has. 

That's fine, and is an expected behavior, but most of the time (and the examples on the webpage shows this as well), the timeset is set with a hard-coded <code>period_in_ns</code> value. This means that there's no way for me, as the simulator (or testers,  if we get to time sheet generation) to know what the desired <code>period_in_ns</code> is.

An easy enough workaround that I've done internally is to just require an <code>attr_reader</code> on the <code>dut</code> called <code>#{timeset_name}_period_in_ns</code>which will be looked up when encountered. For example, to switch to the <code>intram</code> timeset, I'll require an <code>attr_reader</code> called <code>intram_period_in_ns</code> to simulate correctly.

This works fine, but isn't an officially supported thing. Is there currently some sort of API we should be using to declare timeset periods, or is this something we should add? Since a single timeset on the tester can't change without revalidation anyway (the period of <code>intram</code> won't randomly change from 40 ns to 60 ns in the course of a single pattern, I mean), I don't think this gets in the way of anything.

I can check in my workspace and point to some example usage if needed.

The tl;dr of this: How can I/how should I access the desired <code>period_in_ns</code>of a timeset, without having first set that timeset?

Thanks!

**Comment by ginty** [comment]
Until now, timesets have been considered period agnostic, so they can be defined with reference to a variable period, e.g. `w.compare :data, at: "period / 4"`, and yes you supply the actual period at 'instantiation' time.

Maybe we should add the concept of a default period that could be declared during the timeset definition, and then if you call `set_timeset` without a period argument it would use the default period if present, otherwise raise an error.

Would that work?

**Comment by ginty** [comment]
Yeah, I agree with your comments about moving away from defining the period at the point of timeset selection.

Why don't we just give the timeset class a `period_in_ns` attribute, then it could be used like this:

~~~ruby
# To set the default period when defining the timeset
timeset :func do |t|
  t.period_in_ns = 100
end

# To change it (or set it in the first place) later on
dut.timeset(:func).period_in_ns = 80

# We should still keep around the ability to provide a final override at selection time
# for legacy compatibility
dut.set_timeset(:func, 40)

# Though this would be the intended approach and fine if all documentation drops references
# to the above form in preference for this one:
dut.set_timeset(:func)
~~~

**Comment by coreyeng** [comment]
Okay, that's fine too! I think not calling it <code>default</code> implies a bit less dynamic behavior (but maybe that's just me). I'll of course keep around the legacy definition but would throw in a deprecation logger message when used. Thanks!

**Comment by ginty** [comment]
I realized why the legacy definition is the way it is - if you're not going anywhere near a simulation and all you want to do is generate an ATE pattern, then you don't need a full timeset definition, all you need is the name of the timeset and the period (so that Origen can convert time-based delays to cycles).

So, I don't think we should add a deprecation warning to the legacy API and force users to define a timeset before they reference it - it would be an unnecessary overhead in many cases.

**Comment by coreyeng** [comment]
Doesn't that still have the problem where trying to change the timeset period in the pattern for the same timeset will cause inconsistencies? If all you want is ATE pattern generation, the timeset would still look pretty simple and we could make it support options too (if it doesn't already) for something like:

~~~ruby
timeset(:ex, period_in_ns: 40)
~~~

I still think its a good idea to register the timesets beforehand and not set the period during the course of the pattern even for just ATE generation.

**Comment by ginty** [comment]
Has this problem ever actually occurred in practice though?
You could still guard against that by raising an error if the period is changed after it is set.
Overall, I'm still not convinced of the need to make the world update here, though I guess the changes I am about to add which add the app directory and a defined place to declare timesets make it a good time to do it.
Maybe as a compromise, we could gate the deprecation warning firing by whether the app is using the new app dir or not. Which would mean that legacy apps are not bothered by this, but all new ones will adopt the preferred behavior.
